thesis_text/thesis_text.md

```diff
@@ -0,0 +1,551 @@
+
+For a more involved example, let's apply this strategy to the following pseudo bytecode of `for(int i = 0; i < 100; i++) {a = a*a; }; return a;`:
+
+```
+; i=0
+0: OPCODE_LOADCONST 0
+1: OPCODE_STOREVAR i
+
+; i < 100
+2: OPCODE_LOADVAR i
+3: OPCODE_LOADCONST 100
+4: OPCODE_JMPIFBEQ @14
+
+; a = a*a
+5: OPCODE_LOADVAR a
+6: OPCODE_LOADVAR a
+7: OPCODE_MULTIPLY
+8: OPCODE_STOREVAR a
+
+; i++
+9: OPCODE_LOADVAR i
+10: OPCODE_LOADCONST 1
+11: OPCODE_ADD
+12: OPCODE_STOREVAR i
+
+; loop
+13: OPCODE_JMP @2
+
+; return a
+14: OPCODE_LOADVAR a
+15: OPCODE_RET
+```
+
+Thanks to the strategy, for every bytecode offset only one state has to be created. Knowing exactly the stack depth, we can partially evaluate the stack positions to constants
+
+```
+; i=0
+0: stack[0] = 0
+1: vars[i] = stack[0]
+
+; i < 100
+2: stack[0] = vars[i]
+3: stack[1] = 100
+4: if(stack[0]>=stack[1]) goto @14
+
+; a = a*a
+5: stack[0] = vars[a]
+6: stack[1] = vars[a]
+7: stack[0] = stack[0] * stack[1]
+8: vars[a] = stack[0]
+
+; i++
+9: stack[0] = vars[i]
+10: stack[1] = 1
+11: stack[0] = stack[0] + stack[1]
+12: vars[i] = stack[0]
+
+; loop
+13: goto @2
+
+; return a
+14: stack[0] = vars[a]
+15: return stack[0]
+```
+
+As the stack is not used outside this method, it will be completely virtualized. The fact that all stack array accesses use constant allows for aggressive optimization resulting in completely optimizing out the array:
+
+```
+  vars[i] = 0;
+
+condition:
+  if(vars[i]>=100) goto end
+  vars[a] = vars[a]*vars[a]
+  vars[i] = vars[i]+1
+  goto condition
+
+end:
+  return vars[a]
+```
+
+The original control flow of the method was reconstructed from flat bytecode just by unrolling the interpreter loop and merging the instances having the same bytecode offset.
+
+# CLI Component parser
+
+Before being able to execute code, it is necessary to read the code from the assemblies. Prior to starting the work, we expected some open source parsers for this format to exist for various languages, including Java. However, the only alternative stand-alone parser (not a component of a full CLI implementation) we found was [dnlib](https://github.com/0xd4d/dnlib) targeting .NET framework itself. If even for such a popular runtime there are no suitable parsers implemented in Java, we feel that the parser implementation step is an important part to consider in the whole "Building an experimental runtime" picture.
+
+## Analysis
+
+### Design goals
+
+Before we started designing and implementing the parser, we considered what additional constraints have to be put on a parser in order for it to be partial-evaluation friendly. For partial-evaluation friendliness, the key metric is how trivial can every piece of code (that can be called a hot path) be partially evaluated to. While our goal was for the parser to never be called on a hot path, for some possible scenarios including reflection it would be necessary.
```

Comment:
"hot path" se bezne mezi Truffle vyvojari pouziva, presnejsi je asi neco jako "included in partial evaluation" nebo tak podobne. Mozna bys mohl nekde na zacatku definovat, co pro nas bude znamenat "hot path". IMO performance parseru je taky dulezita, coz je jedna z vyhod tve prace oproti te praci z Lince, a tak i v parseru budou nejake "hot paths" v urcitem slova smyslu

Reply:
Premyslim, jestli to nejak nedostat do problem statementu, protoze v jistem smyslu to byl tvuj goal udelat tohle lip nez to mel ten clovek z Lince. Objektivne vzato, pokud by bylo cilem jen udelat rychly interpreter, pak zkraceni cesty k tomuto cili pres zjenoduseni parsovani by mohlo byt validni. Dalsi hezka vec je, ze mas i language launcher podle Truffle "best practices", treba by se to taky dalo prodat i s tim